src/lib/operators/projection.cpp

```diff
@@ -81,6 +82,40 @@ std::shared_ptr<const Table> Projection::_on_execute() {
         output_segments[column_id] = input_chunk->get_segment(pqp_column_expression->column_id);
         column_is_nullable[column_id] =
             column_is_nullable[column_id] || input_table.column_is_nullable(pqp_column_expression->column_id);
+      } else if (expression->type == ExpressionType::PQPColumn && !forward_columns) {
+        // Cannot immediately forward the column because it is a ReferenceSegment and we need to return a ValueSegment.
+        // However, instead of going through the ExpressionEvaluator just to materialize the segment, we can do it
+        // here:
+        const auto pqp_column_expression = std::static_pointer_cast<PQPColumnExpression>(expression);
+        const auto segment = input_chunk->get_segment(pqp_column_expression->column_id);
+
+        resolve_data_type(expression->data_type(), [&](const auto data_type) {
+          using ColumnDataType = typename decltype(data_type)::type;
+          bool has_null = false;
+          auto values = pmr_concurrent_vector<ColumnDataType>(segment->size());
+          auto null_values = pmr_concurrent_vector<bool>(segment->size());
+
+          auto chunk_offset = ChunkOffset{0};
+          segment_iterate<ColumnDataType>(*segment, [&](const auto& position) {
+            if (position.is_null()) {
```

Comment:
Could we get rid of the branch by something like `is_null = pos.is_null();  has_null |= is_null;` and always writing to `values` and `null_values`?

Reply:
Yes, but then we would always be writing to values and null_values. Hard to benchmark as we don't really do too much with NULLs yet.

Reply:
Not really a good argument, but this is consistent with how we do it in other places.

Reply:
Well, we don't do much with NULLs but the branch is there anyway. Worth to benchmark it?

Reply:
on it

Reply:
```diff
 +----------------+----------------+-------+----------------+-------+------------+---------+
 | Benchmark      | prev. iter/s   | runs  | new iter/s     | runs  | change [%] | p-value |
 +----------------+----------------+-------+----------------+-------+------------+---------+
 | TPC-H 01       | 0.614538609982 | 37    | 0.603783071041 | 37    | -2%        |  0.0076 |
 | TPC-H 02       | 66.7094421387  | 4003  | 67.6907730103  | 4062  | +1%        |  0.0571 |
 | TPC-H 03       | 7.26622486115  | 436   | 7.09939622879  | 426   | -2%        |  0.0005 |
 | TPC-H 04       | 6.85076665878  | 412   | 6.92632436752  | 416   | +1%        |  0.0186 |
-| TPC-H 05       | 3.85983896255  | 232   | 3.6632399559   | 220   | -5%        |  0.0000 |
 | TPC-H 06       | 198.411514282  | 11905 | 198.467041016  | 11909 | +0%        |  0.8153 |
 | TPC-H 07       | 7.32884645462  | 440   | 7.52846670151  | 452   | +3%        |  0.0002 |
 | TPC-H 08       | 6.23874807358  | 375   | 6.14935827255  | 369   | -1%        |  0.0229 |
 | TPC-H 09       | 1.53204488754  | 92    | 1.52867507935  | 92    | -0%        |  0.6390 |
 | TPC-H 10       | 2.75670576096  | 166   | 2.69066119194  | 162   | -2%        |  0.0049 |
 | TPC-H 11       | 28.2791900635  | 1697  | 28.844127655   | 1731  | +2%        |  0.0000 |
 | TPC-H 12       | 8.70268154144  | 523   | 8.7239151001   | 524   | +0%        |  0.0020 |
 | TPC-H 13       | 2.21104550362  | 133   | 2.1796412468   | 131   | -1%        |  0.0000 |
 | TPC-H 14       | 46.188533783   | 2772  | 44.4472465515  | 2667  | -4%        |  0.0000 |
 | TPC-H 15       | 68.9120635986  | 4135  | 67.1021194458  | 4027  | -3%        |  0.0000 |
 | TPC-H 16       | 6.9563832283   | 418   | 6.91787004471  | 416   | -1%        |  0.0978 |
 | TPC-H 17       | 4.54050970078  | 273   | 4.64950180054  | 279   | +2%        |  0.0013 |
-| TPC-H 18       | 0.826413154602 | 50    | 0.77485537529  | 47    | -6%        |  0.0000 |
 | TPC-H 19       | 8.58924388885  | 516   | 8.95484256744  | 538   | +4%        |  0.0000 |
 | TPC-H 20       | 25.4696979523  | 1529  | 26.2662220001  | 1576  | +3%        |  0.0000 |
 | TPC-H 21       | 1.14416277409  | 69    | 1.129114151    | 68    | -1%        |  0.2558 |
 | TPC-H 22       | 13.876742363   | 833   | 13.9337511063  | 837   | +0%        |  0.2627 |
 | geometric mean |                |       |                |       | -1%        |         |
 +----------------+----------------+-------+----------------+-------+------------+---------+
```

As the notable regressions are exactly in those queries that were previously improved, this seems to hurt performance.

Reply:
Ok, interesting. So the only situation where we __might__ see a benefit will be 50% of NULLs? Probably never going to happen. I try to remember that before I annoy you with that the next time. 🙂